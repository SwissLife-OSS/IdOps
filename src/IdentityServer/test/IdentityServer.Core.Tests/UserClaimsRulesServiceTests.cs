using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using IdOps.IdentityServer.Model;
using IdOps.IdentityServer.Services;
using IdOps.IdentityServer.Store;
using Microsoft.Extensions.Caching.Memory;
using Moq;
using Snapshooter.Xunit;
using Xunit;

namespace IdOps
{
    public class UserClaimsRulesServiceTests
    {
        [Fact]
        public async Task GetRulesByClient_ByTenantAndClientId_ReturnsSameTenantAndClientIdAndWithNullClient()
        {
            // Arrange
            Mock<IUserClaimRuleRepository> mock = CreateRepositoryMock();

            var service = new UserClaimsRulesService(mock.Object, new MemoryCache(new MemoryCacheOptions()));
            var client = new IdOpsClient
            {
                Tenant = "Tenant_A",
                ClientId = "client_a"
            };

            var claimType = new List<string>();

            // Act
            IReadOnlyList<UserClaimRule> rules = await service.GetRulesByClientAsync(client, claimType, CancellationToken.None);

            // Assert
            rules.MatchSnapshot();
        }

        private Mock<IUserClaimRuleRepository> CreateRepositoryMock()
        {
            var mock = new Mock<IUserClaimRuleRepository>();

            mock.Setup(m => m.GetAllAsync(It.IsAny<CancellationToken>()))
                .ReturnsAsync(new List<UserClaimRule>
                {
                    new UserClaimRule
                    {
                        Name = "a",
                        Tenant = "Tenant_A",
                        ClientIds = new[] {"client_a" },
                    },
                    new UserClaimRule
                    {
                        Name = "a",
                        Tenant = "Tenant_A",
                        ClientIds = new[] {"client_z" },
                    },
                    new UserClaimRule
                    {
                        Name = "b",
                        Tenant = "Tenant_A",
                        ClientIds = Array.Empty<string>()
                    },
                    new UserClaimRule
                    {
                        Name = "c",
                        Tenant = "Tenant_B"
                    },
                    new UserClaimRule
                    {
                        Name = "d",
                        Tenant = "Tenant_B",
                        ClientIds = new[] {"client_d" },
                    }
                });

            return mock;
        }
    }
}
