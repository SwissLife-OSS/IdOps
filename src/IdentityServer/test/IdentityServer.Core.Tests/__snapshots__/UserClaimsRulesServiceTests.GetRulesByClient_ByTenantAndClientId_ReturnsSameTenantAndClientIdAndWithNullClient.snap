[
  {
    "Name": "b",
    "Tenant": "Tenant_A",
    "ClientIds": [],
    "Rules": [],
    "Claims": [],
    "Source": null
  },
  {
    "Name": "c",
    "Tenant": "Tenant_B",
    "ClientIds": [],
    "Rules": [],
    "Claims": [],
    "Source": null
  }
]
